src/index.ts

```diff
@@ -1,259 +1,214 @@
+const log = debug('settlement-xrp')
 
-    // Add redis to context
-    this.app.context.redis = this.redis
-    this.app.context.settlement_prefix = DEFAULT_SETTLEMENT_ENGINE_PREFIX
-    this.app.context.settleAccount = this.settleAccount.bind(this)
-    this.app.context.xrpAddress = this.address
+export const TESTNET_RIPPLED_URI = 'wss://s.altnet.rippletest.net:51233'
 
-    this.router = new Router()
-    this.setupRoutes()
+export interface XrpEngineOpts {
+  xrpSecret?: string
+  rippledUri?: string
+  rippleClient?: RippleAPI
+}
 
-    this.app.use(this.router.routes())
-  }
+export interface XrpSettlementEngine extends SettlementEngine {
+  handleMessage(accountId: string, message: any): Promise<any>
+  handleTransaction(tx: any): void
+  disconnect(): Promise<void>
+}
 
-  public async start () {
-    this.server = this.app.listen(this.port)
-    await this.rippleClient.connect()
-    await this.subscribeToTransactions()
-  }
+export type ConnectXrpSettlementEngine = (services: AccountServices) => Promise<XrpSettlementEngine>
 
-  public async shutdown () {
-    await Promise.all([
-      this.server.close(),
-      this.rippleClient.disconnect()
-    ])
-  }
+export const createEngine = (opts: XrpEngineOpts = {}): ConnectXrpSettlementEngine => async ({
+  sendMessage,
+  creditSettlement
+}) => {
+  const xrpSecret = opts.xrpSecret || (await generateTestnetAccount())
+  const xrpAddress = secretToAddress(xrpSecret)
 
-  private setupRoutes () {
-    this.router.post('/accounts', ctx => createAccount(ctx))
-    this.router.get('/accounts/:id', ctx => showAccount(ctx))
-    this.router.delete('/accounts/:id', ctx => destroyAccount(ctx))
+  const rippleClient: RippleAPI =
+    opts.rippleClient ||
+    new RippleAPI({
+      server: opts.rippledUri || TESTNET_RIPPLED_URI
+    })
 
-    // Account Messages
-    this.router.post('/accounts/:id/messages', this.findAccountMiddleware, createAccountMessage)
+  const incomingPaymentTags = new Map<number, string>() // destinationTag -> accountId
 
-    // Account Settlements
-    this.router.post(
-      '/accounts/:id/settlements',
-      this.findAccountMiddleware,
-      createAccountSettlement
-    )
-  }
+  const self: XrpSettlementEngine = {
+    async handleMessage(accountId, message) {
+      if (message.type && message.type === 'paymentDetails') {
+        const destinationTag = randomBytes(4).readUInt32BE(0)
+        if (incomingPaymentTags.has(destinationTag)) {
+          throw new Error('Failed to generate new destination tag')
+        }
 
-  private async subscribeToTransactions () {
-    this.rippleClient.connection.on('transaction', this.handleTransaction.bind(this))
-    await this.rippleClient.request('subscribe', {
-      accounts: [this.address]
-    })
-  }
+        incomingPaymentTags.set(destinationTag, accountId)
+        setTimeout(() => incomingPaymentTags.delete(destinationTag), 5 * 60000) // Clean-up tags after 5 mins to prevent memory leak
```

Comment:
I don't really understand the logic for doing a round trip request to the peer before every settlement. It seems easier to request these details once during the handshake and keep using them. If we were really worried about the details changing, we could support re-initiating the handshake if your details change

Reply:
1. Provides assurance that the peer is online before sending a payment
   - Alternatively, in case the recipient is offline, there'd need to be functionality to lookup past payments and save the most recent ledger version to a DB, since the recipient currently only listens to a real-time transaction stream. This approach minimizes that error case but is much simpler
2. Enables seamlessly switching XRP accounts
   - Alternatively we'd need to persist to a DB which XRP account we're using between runs, so it'd know whether to tell the peer that there's a new XRP account
   - We'd also need to persist XRP accounts for peers to a DB

Since XRP settlements already take 4 seconds, the added latency seems negligible, especially considering on-ledger settlements should be somewhat infrequent